clippy_lints/src/methods/mod.rs

```diff
@@ -1581,6 +1587,48 @@ impl<'tcx> LateLintPass<'tcx> for Methods {
             }
         }
     }
+
+    fn check_trait_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx TraitItem<'_>) {
+        if_chain! {
+            if !in_external_macro(cx.tcx.sess, item.span);
+            if !item.span.from_expansion();
+            if item.ident.name == sym!(new);
+            if let TraitItemKind::Fn(FnSig { decl, .. }, _) = &item.kind;
+            if let FnRetTy::Return(ret_ty) = &decl.output;
+
+            then {
+                let mut visitor = HasSelfVisitor { has_self_ty: false };
+                visitor.visit_ty(ret_ty);
+                if !visitor.has_self_ty {
+                    span_lint(
+                        cx,
+                        NEW_RET_NO_SELF,
+                        item.span,
+                        "methods called `new` usually return `Self`",
+                    );
+                }
+            }
+        }
+    }
+}
+
+struct HasSelfVisitor {
```

Comment:
why can't we use the same `contains_self_ty` as above?

Reply:
ideally merging the logic of both options into a single method

Reply:
Types are not the same `rustc_hir::Ty` and `rustc_middle::ty::TyS`, but I'll try to find a solution

Reply:
you can get the TyS the same way using `return_ty`

Reply:
In `check_trait_item`, I cannot perform `TyS::same_type` as I can't compare it to any `Self`. I can reach the `DefId` of the trait definition using `Res::SelfTy` but cannot find something to compare with.

Reply:
Ah I see. Hmm

Reply:
There should be an unbound Self type variable but I'm not sure how best to get it. @eddyb is there a way to get the Self for a trait def?

Reply:
Not sure if it's the best way tbh, but one way is to create a dummy `TraitRef` using [TraitRef::identity](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TraitRef.html#method.identity) passing the trait's `DefId`, then calling [self_ty](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TraitRef.html#method.self_ty) on that. Another way would be to call `tcx.generics_of(trait_def_id)` and "manually" transform the first parameter into a `ty::Ty`.

Reply:
Thanks! `TraitRef::identity(...).self_ty()` is exactly what I was looking for